src/main/java/net/minecraftforge/common/crafting/CraftingHelper.java

```diff
@@ -235,4 +253,24 @@ public static ICondition getCondition(JsonObject json)
             throw new JsonSyntaxException("Unknown condition type: " + condition.getID().toString());
         return serializer.getJson(condition);
     }
+
+    @Nullable
+    public static ICondition.IContext buildContext(@Nullable TagManager tagManager) {
+        if (tagManager == null) return null;
+
+        Map<ResourceKey<?>, Map<ResourceLocation, Tag<Holder<?>>>> tagsMap = new IdentityHashMap<>();
```

Comment:
Is making a copy of the tag map needed? Or could the context just return the tag manager itself? Have not looked into what is public on tag manager too much

If the copy is needed, I would suggest lazy loading this so it works with my above suggestion.

Reply:
> Is making a copy of the tag map needed?

It's not a deep copy, it only turns the List<Pair<registry key, tag map>> into a Map<registry key, tag map> to avoid having to iterate over all the elements.
See also:
![image](https://user-images.githubusercontent.com/13494793/161139275-699fb64b-1b8d-4b37-8a29-0a7f15f39e2f.png)
![image](https://user-images.githubusercontent.com/13494793/161139288-5ec2b2e1-2ae2-4d93-b833-8fd454c4b6a1.png)

The idea is to prevent iterating over all load results every time a tag is queried.